gio/src/list_store.rs

```diff
@@ -76,6 +76,45 @@ impl ListStore {
         self.splice(self.n_items(), 0, additions)
     }
 
+    // rustdoc-stripper-ignore-next
+    /// Retains only the elements specified by the predicate.
+    /// This method operates in place, visiting each element exactly once in the original order,
+    /// and preserves the order of the retained elements.
+    /// Because the elements are visited exactly once in the original order,
+    /// external state may be used to decide which elements to keep.
+    ///
+    /// # Panics
+    /// Panics if the predicate closure mutates the list by removing or adding items.
+    pub fn retain(&self, f: impl Fn(&glib::Object) -> bool) {
+        let mut consec_removed = 0;
+        let mut i = 0;
+        const ADDITIONS: &[glib::Object] = &[]; // To satisfy the type checker
+
+        let changed = Rc::new(Cell::new(false));
+        let changed_clone = changed.clone();
+        let signal_id = self.connect_items_changed(move |_list, _, _, _| changed_clone.set(true));
+
+        while i < self.n_items() {
+            let keep = f(self.item(i).unwrap().as_ref());
+            if changed.get() {
+                panic!("The closure passed to ListStore::retain() must not mutate the list store");
+            }
+            if !keep {
+                consec_removed += 1;
+            } else if consec_removed > 0 {
+                self.splice(i - consec_removed, consec_removed, ADDITIONS);
+                changed.set(false);
+                i -= consec_removed;
+                consec_removed = 0;
+            }
+            i += 1;
```

Comment:
I hope the test covers the case of `consec_removed > 0` and afterwards `i += 1` and then checks that the next iteration ends up at the correct item?

That part is a bit hard to follow but otherwise this seems all correct :)

Reply:
Eh, it should be covered. If it weren't, `retain` would remove or keep an unwanted item, but the test case checks that.

Reply:
I guess we should add coverage reports to the CI. That was very useful for these kind of things in gstreamer :) If it's covered, all good.